### PR TITLE
AT-3792 Asset class refactor

### PR DESCRIPTION
**Pull request by andher1802**
Tracked here [AT-3792](https://up42.atlassian.net/browse/AT-3792?atlOrigin=eyJpIjoiNDBmNzdlNDIyN2VjNGFkNGJmNDI3NTFmOGY5M2I5ZGUiLCJwIjoiaiJ9)

Items:
* [ ] Ran test & live-tests
* [ ] Implemented (new) unit tests
* [ ] Removed credentials
* [ ] Removed argument pointing to staging
* [ ] Updated [documentation](sdk.up42.com)

For release:
* [ ] Bumped version
* [ ] Added changelog
* [ ] Updated announcement banner

[AT-3792]: https://up42.atlassian.net/browse/AT-3792?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: up42/asset.py [hunk 1]
```diff
@@ -19,6 +19,8 @@
 LIMIT = 50
 
 
+
+
```
Comment (javidq) [inline up42/asset.py:23]: Sorry for the mess, please clean this up

File: tests/test_asset.py [hunk 2]
```diff
@@ -138,3 +138,48 @@ def test_download_stac_asset(asset_mock2, requests_mock):
...
+        "createdAt": "2023-10-19T12:00:00Z",
+        "size": 1024,
+    }
+    representation = repr(asset_mock)
+    ASSET_ID = "363f89c1-3586-4b14-9a49-03a890c3b593"
+    expected_representation = f"Asset(name: Test Asset, asset_id: {ASSET_ID}, createdAt: 2023-10-19T12:00:00Z, size: 1024)"
+    assert representation == expected_representation
+
+
+def test_asset_repr_with_optional_fields(asset_mock):
```
Comment (javidq) [inline tests/test_asset.py:155]: we need only one test for repr and we need one test for asset info retrieval when asset_id is provided

File: up42/asset.py [hunk 3]
```diff
@@ -30,35 +32,22 @@ class Asset:
     ```
     """
 
+    # don't need to provide asset_id and asset_info at the same time
```
Comment (javidq) [inline up42/asset.py:35]: we need to throw exception in this case and that needs to be covered by tests

**Review by javidq — CHANGES_REQUESTED** [review]

File: tests/test_asset.py [hunk 4]
```diff
@@ -138,3 +138,48 @@ def test_download_stac_asset(asset_mock2, requests_mock):
         )
         assert out_path.exists()
         assert out_path.name == "bsg-104-20230522-044750-90756881_ortho.tiff"
+
+
+def test_asset_repr_without_optional_fields(asset_mock):
```
[the remark this answers is not in the document]
Comment (andher1802) [reply tests/test_asset.py]: This one I am not sure if I got it. I am using the asset_mock to retrieve an asset without calling the actual endpoint. Then, I assert the representation with the expected value (The info representation). But I am not sure why shouldn't I use the mock.

File: tests/test_asset.py [hunk 5]
```diff
@@ -133,8 +149,13 @@ def test_download_stac_asset(asset_mock2, requests_mock):
...
+        out_path = asset_mock_2.download_stac_asset(
             pystac.Asset(href=STAC_ASSET_HREF, roles=["data"]), tempdir
         )
         assert out_path.exists()
         assert out_path.name == "bsg-104-20230522-044750-90756881_ortho.tiff"
+
+
+def test_asset_repr(asset_mock):
+    representation = repr(asset_mock)
+    assert representation == repr(asset_mock.info)
```
Comment (seedlit) [inline tests/test_asset.py:161]: interesting! :) 
If I am not wrong, `repr` returns a printable representation of the give object right? 
What are we testing here?

File: up42/asset.py [hunk 6]
```diff
@@ -30,35 +30,30 @@ class Asset:
...
-        representation = (
-            f"Asset(name: {self._info['name']}, asset_id: {self.asset_id}, createdAt: {self._info['createdAt']}, "
-            f"size: {self._info['size']})"
-        )
-        if "source" in self._info:
-            representation += f", source: {self._info['source']}"
-        if "contentType" in self._info:
-            representation += f", contentType: {self._info['contentType']}"
-        return representation
+        return f"{self.info.__repr__()}"
```
Comment (seedlit) [inline up42/asset.py:48]: nice!

**Review by seedlit — APPROVED** [review]
left a non-blocking question :)

File: tests/test_asset.py (hunk 5 above)
Comment (andher1802) [reply tests/test_asset.py:161]: We simplified the repr of the object to be the printed version of the info property (which is a dict). Therefore, we tested that repr has the expected value. Also, increase the coverage. :)

File: up42/asset.py [hunk 7]
```diff
@@ -30,35 +30,30 @@ class Asset:
     ```
     """
 
-    def __init__(self, auth: Auth, asset_id: str, asset_info: Optional[dict] = None):
+    def __init__(
+        self,
+        auth: Auth,
+        asset_id: Optional[str] = None,
```
Comment (javidq) [inline up42/asset.py:36]: str | None

File: up42/asset.py [hunk 8]
```diff
@@ -30,35 +30,30 @@ class Asset:
...
+        asset_info: Optional[dict] = None,
```
Comment (javidq) [inline up42/asset.py:37]: dict | None

File: tests/test_asset.py [hunk 9]
```diff
@@ -25,6 +25,22 @@ def test_init(asset_mock):
     assert asset_mock.asset_id == ASSET_ID
 
 
+def test_init_asset_failure(auth_mock):
```
Comment (javidq) [inline tests/test_asset.py:28]: Let's decompose this into 2 simple separate scenarios:
1. test_init_should_fail_if_both_id_and_info_given
2. test_init_should_fail_if_neither_id_nor_info_given

File: tests/test_asset.py [hunk 10]
```diff
@@ -25,6 +25,22 @@ def test_init(asset_mock):
...
+    with pytest.raises(ValueError) as e:
+        failed_asset = Asset(
```
Comment (javidq) [inline tests/test_asset.py:30]: the variable is not needed

File: tests/test_asset.py [hunk 11]
```diff
@@ -25,6 +25,22 @@ def test_init(asset_mock):
...
+            auth_mock, asset_id="some_asset_id", asset_info={"id": "some_asset_id"}
+        )
+    assert "cannot be provided simultaneously." in str(e.value)
```
Comment (javidq) [inline tests/test_asset.py:33]: Not needed

Comment (javidq) [reply tests/test_asset.py:33]: We're not interested in testing the static log messages

File: tests/test_asset.py [hunk 12]
```diff
@@ -25,6 +25,22 @@ def test_init(asset_mock):
...
+
+    with pytest.raises(ValueError) as e:
+        failed_asset = Asset(auth_mock)
+    assert "Either asset_id or asset_info should be provided" in str(e.value)
+
+
+def test_asset_id(asset_mock):
```
Comment (javidq) [inline tests/test_asset.py:40]: Let's rewrite this without asset_mock and rename into `test_should_get_id_from_info`

File: tests/test_asset.py [hunk 13]
```diff
@@ -51,12 +67,12 @@ def test_asset_update_metadata(asset_mock):
     "asset_fixture, download_url",
     [
         ("asset_mock", DOWNLOAD_URL),
-        ("asset_mock2", DOWNLOAD_URL2),
+        ("asset_mock_2", DOWNLOAD_URL2),
     ],
 )
 def test_asset_get_download_url(asset_fixture, download_url, request):
```
Comment (javidq) [inline tests/test_asset.py:73]: The method is internal, shall not be tested explicitly, we will drop this, so let's avoid changing it

File: tests/test_asset.py [hunk 14]
```diff
@@ -133,8 +149,13 @@ def test_download_stac_asset(asset_mock2, requests_mock):
...
     with tempfile.TemporaryDirectory() as tempdir:
-        out_path = asset_mock2.download_stac_asset(
+        out_path = asset_mock_2.download_stac_asset(
             pystac.Asset(href=STAC_ASSET_HREF, roles=["data"]), tempdir
         )
         assert out_path.exists()
         assert out_path.name == "bsg-104-20230522-044750-90756881_ortho.tiff"
+
+
+def test_asset_repr(asset_mock):
```
Comment (javidq) [inline tests/test_asset.py:159]: 1. test_should_be_represented_by_info
2. no asset_mock, please use only auth_mock

File: tests/test_asset.py [hunk 15]
```diff
@@ -133,8 +149,13 @@ def test_download_stac_asset(asset_mock2, requests_mock):
...
+    representation = repr(asset_mock)
```
Comment (javidq) [inline tests/test_asset.py:160]: Let's inline this variable

**Review by javidq — CHANGES_REQUESTED** [review]

File: tests/test_asset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (javidq) [reply tests/test_asset.py]: We will cover the reasons for that later, the main problem that the way they are implemented they introduce global coupling between testing modules and also hurt a lot local readability in the tests.

File: up42/asset.py (hunk 7 above)
Comment (seedlit) [reply up42/asset.py:36]: > str | None

this is not supported in Python 3.8
SDK has to support python 3.8

File: up42/asset.py [hunk 16]
```diff
@@ -30,35 +32,22 @@ class Asset:
...
     @property
-    def info(self) -> dict:
-        """
-        Gets and updates the asset metadata information.
-        """
-        url = f"{self.auth._endpoint()}/v2/assets/{self.asset_id}/metadata"
-        response_json = self.auth._request(request_type="GET", url=url)
-        self._info = response_json
-        return self._info
+    def asset_id(self) -> dict:
```
[the remark this answers is not in the document]
Comment (andher1802) [reply up42/asset.py:50]: I have added the `test_should_get_id_from_info`

File: tests/test_asset.py (hunk 15 above)
Comment (andher1802) [reply tests/test_asset.py]: I have redone this test to use only `auth_mock` and `asset_info`

File: tests/test_asset.py [hunk 17]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
     assert asset_mock.asset_id == ASSET_ID
 
 
+def test_repr(auth_mock):
+    asset_info = {"id": ASSET_ID}
+    asset = Asset(auth_mock, asset_info=asset_info)
+    expected_repr = "{'id': '363f89c1-3586-4b14-9a49-03a890c3b593'}"
```
Comment (javidq) [inline tests/test_asset.py:40]: expected_repr = repr(asset_info)

File: tests/test_asset.py [hunk 18]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
     assert asset_mock.asset_id == ASSET_ID
 
 
+def test_repr(auth_mock):
+    asset_info = {"id": ASSET_ID}
```
Comment (javidq) [inline tests/test_asset.py:38]: {"id": ASSET_ID, "other": "data"}

File: tests/test_asset.py [hunk 19]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
     assert asset_mock.asset_id == ASSET_ID
 
 
+def test_repr(auth_mock):
```
Comment (javidq) [inline tests/test_asset.py:37]: test_should_delegate_repr_to_info

File: tests/test_asset.py [hunk 20]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
...
+    asset_info = {"id": ASSET_ID}
+    asset = Asset(auth_mock, asset_info=asset_info)
```
Comment (javidq) [inline tests/test_asset.py:39]: from unittest.mock import MagicMock
unreachable=MagicMock()

File: tests/test_asset.py [hunk 21]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
...
+            {"id": ASSET_ID},
+            "asset_id and asset_info cannot be provided simultaneously.",
+        ),
+    ],
+    ids=[
+        "Sc 1: Both asset_id and asset_info provided",
+        "Sc 2: Both asset_id and asset_info not provided",
+    ],
+)
+def test_init_asset_failure(auth_mock, asset_id, asset_info, expected_error):
```
Comment (javidq) [inline tests/test_asset.py:63]: test_init_should_accept_only_asset_id_or_info

File: tests/test_asset.py [hunk 22]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
...
+    with pytest.raises(ValueError) as err:
+        Asset(auth_mock, asset_id=asset_id, asset_info=asset_info)
```
Comment (javidq) [inline tests/test_asset.py:65]: unreachable mock

File: tests/test_asset.py [hunk 23]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
...
+        assert expected_error == err
```
Comment (javidq) [inline tests/test_asset.py:66]: This code is unreachable

File: tests/test_asset.py [hunk 24]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
...
+
+
+def test_should_get_id_from_info(auth_mock, requests_mock):
+    url_asset_info = f"{auth_mock._endpoint()}/v2/assets/{ASSET_ID}/metadata"
+    requests_mock.get(url=url_asset_info, json=JSON_ASSET)
```
Comment (javidq) [inline tests/test_asset.py:71]: These lines are not needed

File: up42/asset.py [hunk 25]
```diff
@@ -25,35 +25,30 @@ class Asset:
...
+        asset_id: Optional[str] = None,
+        asset_info: Optional[dict] = None,
+    ):
+        if asset_id is not None and asset_info is not None:
+            raise ValueError("asset_id and asset_info cannot be provided simultaneously.")
+        if asset_id is None and asset_info is None:
+            raise ValueError("Either asset_id or asset_info should be provided in the constructor.")
         self.auth = auth
-        self.asset_id = asset_id
+        self.info = asset_info or self._get_info(asset_id)
```
Comment (javidq) [inline up42/asset.py:39]: This is not covered

File: tests/test_asset.py [hunk 26]
```diff
@@ -30,6 +34,45 @@ def test_init(asset_mock):
...
+    asset = Asset(auth=auth_mock, asset_info={"id": ASSET_ID})
+    assert asset.asset_id == ASSET_ID
+
+
```
Comment (javidq) [inline tests/test_asset.py:75]: test1: test_should_initialize_with_provided_info - pass asset_info, unreachable auth_mock, verify info
test2: test_should_initialize_with_retrieved_info - pass asset_id, mock response data and verify the call and the result

**Review by javidq — CHANGES_REQUESTED** [review]

File: tests/test_asset.py (hunk 20 above)
Comment (andher1802) [reply tests/test_asset.py:40]: I replaced the auth_mock with the MagicMock, I guess to isolate further the test given that auth_mock is not required to check if asset_info was correctly assigned. Please let's check if that was the idea for this suggestion.

File: tests/test_asset.py (hunk 22 above)
Comment (andher1802) [reply tests/test_asset.py:65]: same here

File: tests/test_asset.py (hunk 26 above)
Comment (andher1802) [reply tests/test_asset.py:75]: 2 options same test with parametrize or 2 tests. I chose the first one. Please double-check any drawbacks.

File: up42/asset.py (hunk 25 above)
Comment (andher1802) [reply up42/asset.py:41]: I think this is covered by adding the `test_should_initialize_with_provided_info`, and `test_should_initialize_with_retrieved_info`

File: up42/storage.py [hunk 27]
```diff
@@ -166,7 +166,7 @@ def get_assets(
         if return_json:
             return assets_json
         else:
-            return [Asset(self.auth, asset_id=asset_json["id"], asset_info=asset_json) for asset_json in assets_json]
+            return [Asset(self.auth, asset_info=asset_json) for asset_json in assets_json]
```
Comment (andher1802) [inline up42/storage.py:169]: required to not provide asset_id and asset_info simultaneously

File: up42/order.py [hunk 28]
```diff
@@ -92,9 +92,7 @@ def get_assets(self) -> List[Asset]:
         if self.is_fulfilled:
             params: AssetSearchParams = {"search": self.order_id}
             assets_response = search_assets(self.auth, params=params)
-            return [
-                Asset(self.auth, asset_id=asset_info["id"], asset_info=asset_info) for asset_info in assets_response
-            ]
+            return [Asset(self.auth, asset_info=asset_info) for asset_info in assets_response]
```
Comment (andher1802) [inline up42/order.py:95]: required to not provide asset_id and asset_info simultaneously

File: tests/test_asset.py [hunk 29]
```diff
@@ -22,14 +22,86 @@
...
 def test_init(asset_mock):
     assert isinstance(asset_mock, Asset)
     assert asset_mock.asset_id == ASSET_ID
 
 
+def test_should_delegate_repr_to_info():
+    asset_info = {"id": ASSET_ID, "other": "data"}
+    asset = Asset(MagicMock(), asset_info=asset_info)
```
Comment (javidq) [inline tests/test_asset.py:46]: auth=MagicMock()

File: tests/test_asset.py [hunk 30]
```diff
@@ -22,14 +22,86 @@
...
+        ),
+    ],
+    ids=[
+        "Sc 1: Both asset_id and asset_info provided",
+        "Sc 2: Both asset_id and asset_info not provided",
+    ],
+)
+def test_init_should_accept_only_asset_id_or_info(asset_id, asset_info, expected_error):
+    with pytest.raises(ValueError) as err:
+        Asset(MagicMock(), asset_id=asset_id, asset_info=asset_info)
```
Comment (javidq) [inline tests/test_asset.py:71]: auth=MagicMock()

**Review by javidq — CHANGES_REQUESTED** [review]

File: tests/test_asset.py (hunk 26 above)
Comment (andher1802) [reply tests/test_asset.py:88]: splitted

File: tests/test_asset.py [hunk 31]
```diff
@@ -22,14 +22,70 @@
...
+    password_test_live,
+    project_api_key_live,
+    project_id_live,
+    storage_live,
+    storage_mock,
+    username_test_live,
 )
 
 
+@pytest.fixture
```
Comment (javidq) [inline tests/test_asset.py:34]: Not needed

File: tests/test_asset.py [hunk 32]
```diff
@@ -22,14 +22,70 @@
...
+    with pytest.raises(ValueError) as err:
+        Asset(auth=MagicMock(), asset_id=asset_id, asset_info=asset_info)
+    assert expected_error == str(err.value)
+
+
+def test_should_initialize_with_retrieved_info(requests_mock, auth_mock):
+    url_asset_info = endpoint(f"/v2/assets/{ASSET_ID}/metadata")
+    requests_mock.get(url=url_asset_info, json=JSON_ASSET)
+    asset = Asset(auth=auth_mock, asset_id=ASSET_ID)
+    assert asset.asset_id == ASSET_ID
```
Comment (javidq) [inline tests/test_asset.py:79]: assert asset.info == JSON_ASSET

File: tests/test_asset.py [hunk 33]
```diff
@@ -22,14 +22,70 @@
...
+
+
+def test_should_initialize_with_provided_info():
+    provided_info = {"id": ASSET_ID, "name": "test name"}
+    asset = Asset(auth=MagicMock(), asset_info=provided_info)
+    assert asset.asset_id == ASSET_ID
+    assert repr(asset.info) == repr(provided_info)
```
Comment (javidq) [inline tests/test_asset.py:86]: assert asset.info == provided_info

**Review by javidq — CHANGES_REQUESTED** [review]

File: tests/test_asset.py (hunk 31 above)
Comment (andher1802) [reply tests/test_asset.py]: I forgot :)

**Review by javidq — APPROVED** [review]

**Pull request merged**